libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,269 @@
+
+    1. Storing the content
+    2. Allowing retrieval of the content by id
+    3. Supporting queries against the metadata associated with the content
+
+    The index is **NOT** responsible for:
+
+    1. Support search queries against the content.
+
+    The types of supported queries seems very dependent on the actual
+    """
+
+    @abc.abstractmethod
+    def upsert(self, data: Iterable[T], /, **kwargs: Any) -> Iterable[UpsertResponse]:
+        """Upsert a stream of data by id."""
+
+    @abc.abstractmethod
+    def get_by_ids(
+        self,
+        ids: Iterable[str],  # Sequence or Iterable
+        /,
+    ) -> Iterable[T]:
+        """Get items by id."""
+
+    @abc.abstractmethod
+    def delete_by_ids(
+        self,
+        ids: Sequence[str],  # Sequence or Iterable
+        /,
+        **kwargs: Any,
+    ) -> DeleteResponse:
+        """Delete an item by id."""
+
+    # Delete and get are part of the READ/WRITE interface.
+    # They do not take advantage of indexes on the content.
+    # However, all the indexers ARE assumed to have the capability to index
+    # on metadata if they implement the delete_by_query and get_by_query methods.
+    @abc.abstractmethod
+    def lazy_get(self, query: FilterQuery, /, **kwargs: Any) -> Iterable[T]:
+        """Get items by query."""
+
+    @abc.abstractmethod
+    def lazy_delete(
+        self, query: FilterQuery, /, **kwargs: Any
+    ) -> Iterable[DeleteResponse]:
+        """Delete items by query."""
+        # Careful with halloween problem
+
+
+class Hit(TypedDict):
+    # Should subclass from content?
+    id: str
+    score: float
+    # Optional source
+    source: Optional[T]
+    metadata: Dict[str, Any]
+
+
+class QueryResponse(TypedDict):
+    """A retrieval result."""
+
+    # Free form metadata for vectorstore providers
+    metadata: Dict[str, Any]
+    hits: List[Hit]
+
+
+class SearchQuery(FilterQuery, Generic[T]):
+    """A query that searches for items according to some criteria.
+
+    This query format is specifically adapted to vectorstores.
+
+    It supports similarity searches together with standard filtering operations.
+
+    We would need to verify that we can accommodate all the existing functionality
+    with it.
+    """
+
+    query: Optional[T]
+    # Optionally specify the method to use for similarity search
+    method: NotRequired[Optional[str]]  # e.g., "cosine", "euclidean", "jaccard", "mmr"
+
+
+class UpsertData(TypedDict):
+    """Data to upsert."""
+
+    id: str
+    document: Document
+    vector: Vector
+
+
+class VectorStore(BaseIndex[Document]):
+    @abc.abstractmethod
+    def query(
+        self,
+        query: SearchQuery[T],
+        /,
+        *,
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Any,
+    ) -> QueryResponse[
+        Document
+    ]:  # kwargs for things that are not captured via the standard query?
+        """Query for items."""
+
+    @abc.abstractmethod
+    def upsert_with_vector(
```

Comment:
It's nice to be able to provide additional data during indexing. For example, for a vectorstore being able to specify the vector is a very obvious thing. What's a good way to represent this while still keeping the upsert interface?